### PR TITLE
Add client support for  batch processing

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/104428

This PR:
* add the `/batch` and `/state` endpoints to the client to support batch push of data
* it still does not remove the polling logic from the client; that needs to be done **after** the service changes
* adds the `aggregatorv1` (functional) test suite
* de-duplicates a lot of weird namespacing in our tests (thus the file "noise")

For your consideration:
* [Main](https://github.com/moderntribe/the-events-calendar/pull/1944/files#diff-e5ad0809a1399fb28c50f3775e36ba7e)
* [Service Provider](https://github.com/moderntribe/the-events-calendar/pull/1944/files#diff-36e3e5516a7484cbcead8328f8907763)
* [the classes handling the requests](https://github.com/moderntribe/the-events-calendar/pull/1944/files#diff-e29c146eebcb36f0dcafb20fb6a2b567) 
* [the tests](https://github.com/moderntribe/the-events-calendar/pull/1944/files#diff-242c4af5fa7e57c6099b4bef47444ba7) 

Check the questions, in form of todos, too!

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;should we check for a valid license here? Usage?&quot; | [Service_Provider.php:23](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L23)
:warning: | Comment refers to a TODO task &quot;should we check the specific length of the import ID here?&quot; | [Service_Provider.php:42](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L42)
:x: | Expected 1 space between argument &quot;$args&quot; and equals sign; 0 found | [Records.php:664](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L664)
:x: | Comma required after last value in array declaration | [Records.php:691](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L691)
:x: | Comma required after last value in array declaration | [Records.php:697](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L697)
:x: | Comma required after last value in array declaration | [Records.php:703](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L703)
:x: | Comma required after last value in array declaration | [Records.php:709](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L709)
:x: | Comma required after last value in array declaration | [Records.php:715](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L715)
:x: | Comma required after last value in array declaration | [Records.php:721](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L721)
:x: | Comma required after last value in array declaration | [Records.php:727](https://github.com/moderntribe/the-events-calendar/blob/c8117f2e6f0c59b53082280c4a972d60c53ec739/src/Tribe/Aggregator/Records.php#L727)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Closing this Pull Request because it is pull requested against the wrong branch. The Central ticket is for *1.0.20*, but this was pull requested against *release/M18.07*.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;should we check for a valid license here? Usage?&quot; | [Service_Provider.php:23](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L23)
:warning: | Comment refers to a TODO task &quot;should we check the specific length of the import ID here?&quot; | [Service_Provider.php:42](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L42)
:x: | Expected 1 space between argument &quot;$args&quot; and equals sign; 0 found | [Records.php:664](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L664)
:x: | Comma required after last value in array declaration | [Records.php:691](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L691)
:x: | Comma required after last value in array declaration | [Records.php:697](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L697)
:x: | Comma required after last value in array declaration | [Records.php:703](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L703)
:x: | Comma required after last value in array declaration | [Records.php:709](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L709)
:x: | Comma required after last value in array declaration | [Records.php:715](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L715)
:x: | Comma required after last value in array declaration | [Records.php:721](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L721)
:x: | Comma required after last value in array declaration | [Records.php:727](https://github.com/moderntribe/the-events-calendar/blob/04ce1f6b35c09023b5f30c98a09a097fa0d006f2/src/Tribe/Aggregator/Records.php#L727)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/e0760ac5da18ceba8023f38bd5c5eb181def6f1c/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/e0760ac5da18ceba8023f38bd5c5eb181def6f1c/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;should we check for a valid license here? Usage?&quot; | [Service_Provider.php:23](https://github.com/moderntribe/the-events-calendar/blob/e0760ac5da18ceba8023f38bd5c5eb181def6f1c/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L23)
:warning: | Comment refers to a TODO task &quot;should we check the specific length of the import ID here?&quot; | [Service_Provider.php:42](https://github.com/moderntribe/the-events-calendar/blob/e0760ac5da18ceba8023f38bd5c5eb181def6f1c/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L42)
:x: | Expected 1 space between argument &quot;$args&quot; and equals sign; 0 found | [Records.php:664](https://github.com/moderntribe/the-events-calendar/blob/e0760ac5da18ceba8023f38bd5c5eb181def6f1c/src/Tribe/Aggregator/Records.php#L664)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/a77cdb0607e503dcd56e3521fe571d992c9ccd57/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/a77cdb0607e503dcd56e3521fe571d992c9ccd57/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;should we check for a valid license here? Usage?&quot; | [Service_Provider.php:23](https://github.com/moderntribe/the-events-calendar/blob/a77cdb0607e503dcd56e3521fe571d992c9ccd57/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L23)
:warning: | Comment refers to a TODO task &quot;should we check the specific length of the import ID here?&quot; | [Service_Provider.php:42](https://github.com/moderntribe/the-events-calendar/blob/a77cdb0607e503dcd56e3521fe571d992c9ccd57/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L42)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+				'validate_callback' => array( $this, 'is_expected_batch_hash' ),
+				'description' => __( 'The hash of the next expected batch, as previously provided by the client', 'the-events-calendar' ),
+			),
+			'status' => array(
+				'required' => true,
+				'type' => 'object',
+				'validate_callback' => array( $this, 'is_valid_status_information' ),
+				'description' => __( 'The current status of the import.', 'the-events-calendar' ),
+			),
+			'done' => array(
```
Comment (borkweb) [inline src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php]: `done` seems like a pretty generic word that could hold a number _or_ a boolean. I wonder if changing it to something more explicit would be useful? Like `percentage_complete`, perhaps?

File: src/Tribe/Aggregator/REST/V1/Endpoints/State.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+				'required' => true,
+				'type' => 'string',
+				'description' => __( 'The new status message for the user, not localized.', 'the-events-calendar' ),
+			),
+			'message_slug' => array(
+				'required' => true,
+				'type' => 'string',
+				'description' => __( 'The new status message slug, to allow for localized messages.', 'the-events-calendar' ),
+			),
+
```
Comment (borkweb) [inline src/Tribe/Aggregator/REST/V1/Endpoints/State.php]: Remove the extra line break?

File: src/Tribe/Aggregator/REST/V1/Endpoints/State.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		/** @var Tribe__Events__Aggregator__Records $records */
+		$records = tribe( 'events-aggregator.records' );
+
+		/** @var Tribe__Events__Aggregator__Record__Abstract $record */
+		$record = $records->get_by_import_id( $request['import_id'], array( 'post_status' => 'any' ) );
+
+		$status = $request['status'];
+
+		$updated = $record->set_status( $status );
+		$record->update_meta( 'done', $request['done'] );
```
Comment (borkweb) [inline src/Tribe/Aggregator/REST/V1/Endpoints/State.php]: Should this line be executed if `$updated` is an error state (empty or WP_Error)? Or should it be pushed down to be executed around line 35-ish?

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/79553802fb8c4e035d024ed83899a75b64d478a4/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/79553802fb8c4e035d024ed83899a75b64d478a4/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;should we check for a valid license here? Usage?&quot; | [Service_Provider.php:23](https://github.com/moderntribe/the-events-calendar/blob/79553802fb8c4e035d024ed83899a75b64d478a4/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L23)
:warning: | Comment refers to a TODO task &quot;should we check the specific length of the import ID here?&quot; | [Service_Provider.php:42](https://github.com/moderntribe/the-events-calendar/blob/79553802fb8c4e035d024ed83899a75b64d478a4/src/Tribe/Aggregator/REST/V1/Service_Provider.php#L42)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
Fixed stuff and ready for another look.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/daf5fedc029c16efc6a8815879888efc0856f146/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/daf5fedc029c16efc6a8815879888efc0856f146/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/e165b033910817452c86806595e1f4884deef1fb/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/e165b033910817452c86806595e1f4884deef1fb/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;let's revisit this to return when more UI is exposed&quot; | [Abstract.php:1203](https://github.com/moderntribe/the-events-calendar/blob/e165b033910817452c86806595e1f4884deef1fb/src/Tribe/Aggregator/Record/Abstract.php#L1203)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/d44f13d5cd52de41078e65d177b328d57ab6377b/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/d44f13d5cd52de41078e65d177b328d57ab6377b/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;let's revisit this to return when more UI is exposed&quot; | [Abstract.php:1221](https://github.com/moderntribe/the-events-calendar/blob/d44f13d5cd52de41078e65d177b328d57ab6377b/src/Tribe/Aggregator/Record/Abstract.php#L1221)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
I've refined the logic to make it so that, for the time being, batch push will not be used but is fully supported.
Essentially we look up a `batch_push` flag in the Service response when queueing the import (`Record__Abstract::queue_import`) and from there know if the record should be polling or wait for pushes.

File: src/Tribe/Aggregator/Tabs/New.php [hunk 4]
```diff
@@ -236,6 +237,13 @@ public function handle_import_finalize( $data ) {
 		}
 	}
 
+	/**
+     * Parses the queue for errors and informations.
```
Comment (bordoni) [inline src/Tribe/Aggregator/Tabs/New.php]: Indentation.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $return_id is never used | [Batch.php:44](https://github.com/moderntribe/the-events-calendar/blob/5259bbb2b891b9ec3a8c636da092190ce531439a/src/Tribe/Aggregator/REST/V1/Endpoints/Batch.php#L44)
:warning: | The method parameter $return_id is never used | [State.php:15](https://github.com/moderntribe/the-events-calendar/blob/5259bbb2b891b9ec3a8c636da092190ce531439a/src/Tribe/Aggregator/REST/V1/Endpoints/State.php#L15)
:warning: | Comment refers to a TODO task &quot;let's revisit this to return when more UI is exposed&quot; | [Abstract.php:1221](https://github.com/moderntribe/the-events-calendar/blob/5259bbb2b891b9ec3a8c636da092190ce531439a/src/Tribe/Aggregator/Record/Abstract.php#L1221)

> via jenkins/codesniffer-phpcs

**Pull request merged**